src/coding_theory/hamming.lean

```diff
@@ -0,0 +1,245 @@
+  by_contra' H, exact h (eq.trans H.1 H.2)
+end
+
+lemma eq_of_hamm_dist_eq_zero (x y : Π i, β i) : hamm_dist x y = 0 → x = y :=
+hamm_dist_eq_zero.mp
+
+lemma hamm_dist_ne_zero {x y : Π i, β i} : hamm_dist x y ≠ 0 ↔ x ≠ y :=
+not_iff_not.mpr hamm_dist_eq_zero
+
+lemma hamm_dist_pos {x y : Π i, β i} : 0 < hamm_dist x y ↔ x ≠ y :=
+by rw [←hamm_dist_ne_zero, iff_not_comm, not_lt, nat.le_zero_iff]
+
+@[simp] lemma hamm_dist_lt_one {x y : Π i, β i} : hamm_dist x y < 1 ↔ x = y :=
+by {rw nat.lt_one_iff, exact hamm_dist_eq_zero}
+
+lemma hamm_dist_eq_zero_iff_forall_eq {x y : Π i, β i} :
+  hamm_dist x y = 0 ↔ ∀ i, x i = y i :=
+by rw [hamm_dist_eq_zero, function.funext_iff]
+
+lemma hamm_dist_ne_zero_iff_exists_ne {x y : Π i, β i} :
+  hamm_dist x y ≠ 0 ↔ ∃ i, x i ≠ y i :=
+by rw [hamm_dist_ne_zero, function.ne_iff]
+
+section has_zero
+
+variables [Π i, has_zero (β i)] [Π i, has_zero (γ i)]
+
+/-- The Hamming weight function to the naturals. -/
+
+def hamm_wt (x : Π i, β i) : ℕ := hamm_dist x 0
+
+lemma hamm_wt_eq (x : Π i, β i) : hamm_wt x = card {i // x i ≠ 0} := rfl
+
+lemma hamm_wt_eq_hamm_dist_zero (x : Π i, β i) : hamm_wt x = hamm_dist x 0 := rfl
+
+lemma hamm_wt_nonneg {x : Π i, β i} : 0 ≤ hamm_wt x := hamm_dist_nonneg
+
+lemma hamm_wt_le_card_fintype {x : Π i, β i} : hamm_wt x ≤ card ι := hamm_dist_le_card_fintype
+
+lemma hamm_wt_comp_le_hamm_wt (f : Π i, γ i → β i) {x : Π i, γ i} (hf : Π i, f i 0 = 0) :
+  hamm_wt (λ i, f i (x i)) ≤ hamm_wt x :=
+by {  simp_rw hamm_wt_eq_hamm_dist_zero, convert hamm_dist_comp_le_hamm_dist f,
+      simp_rw [pi.zero_apply, hf], refl }
+
+lemma hamm_wt_comp (f : Π i, γ i → β i) {x : Π i, γ i} (hf₁ : Π i, injective (f i))
+  (hf₂ : Π i, f i 0 = 0) : hamm_wt (λ i, f i (x i)) = hamm_wt x :=
+by {  simp_rw hamm_wt_eq_hamm_dist_zero, convert hamm_dist_comp f hf₁,
+      simp_rw [pi.zero_apply, hf₂], refl }
+
+lemma hamm_wt_smul_le_hamm_wt [has_zero α] [Π i, smul_with_zero α (β i)] {k : α}
+  {x : Π i, β i} : hamm_wt (k • x) ≤ hamm_wt x :=
+hamm_wt_comp_le_hamm_wt (λ i (c : β i), k • c) (λ i, by simp_rw smul_zero')
+
+lemma hamm_wt_smul [has_zero α] [Π i, smul_with_zero α (β i)] {k : α}
+  (hk : ∀ i, is_smul_regular (β i) k) (x : Π i, β i) : hamm_wt (k • x) = hamm_wt x :=
+hamm_wt_comp (λ i (c : β i), k • c) hk (λ i, by simp_rw smul_zero')
+
+@[simp] lemma hamm_wt_eq_zero {x : Π i, β i} : hamm_wt x = 0 ↔ x = 0 := hamm_dist_eq_zero
+
+@[simp] lemma hamm_wt_zero : hamm_wt (0 : Π i, β i) = 0 := hamm_dist_self _
+
+lemma hamm_wt_ne_zero {x : Π i, β i} : hamm_wt x ≠ 0 ↔ x ≠ 0 := hamm_dist_ne_zero
+
+lemma hamm_wt_pos {x : Π i, β i} : 0 < hamm_wt x ↔ x ≠ 0 := hamm_dist_pos
+
+@[simp] lemma hamm_wt_lt_one {x : Π i, β i} : hamm_wt x < 1 ↔ x = 0 := hamm_dist_lt_one
+
+lemma hamm_wt_zero_iff_forall_zero {x : Π i, β i} : hamm_wt x = 0 ↔ ∀ i, x i = 0 :=
+hamm_dist_eq_zero_iff_forall_eq
+
+lemma hamm_wt_pos_iff_exists_nz {x : Π i, β i} : hamm_wt x ≠ 0 ↔ ∃ i, x i ≠ 0 :=
```

Comment:
```suggestion
lemma hamm_wt_pos_iff_exists_nonzero {x : Π i, β i} : hamm_wt x ≠ 0 ↔ ∃ i, x i ≠ 0 :=
```
We don't use `_nz` anywhere else in lemma names it seems.

Reply:
Should be `ne_zero` I think actually.